pallets/identity/src/keys.rs

```diff
@@ -214,36 +237,69 @@ impl<T: Config> Module<T> {
                 );
                 // Ensure authorizations are for the same DID.
                 ensure!(
-                    rotation_for_did == attestation_for_did,
+                    target_did == attestation_for_did,
                     Error::<T>::AuthorizationsNotForSameDids
                 );
                 Ok(())
             })?;
         }
 
-        Self::ensure_key_did_unlinked(&sender)?;
-
-        // Get the current DidRecord.
-        let mut record = Self::did_records(&rotation_for_did);
-        let old_primary_key = record.primary_key;
-
-        // Ensure that it is safe to unlink the primary key from the did.
-        Self::ensure_key_unlinkable_from_did(&old_primary_key)?;
+        // Remove old secondary key
+        if is_secondary_key {
+            let removed_signers = vec![signer];
+            record.remove_secondary_keys(&removed_signers);
+            Self::deposit_event(RawEvent::SecondaryKeysRemoved(target_did, removed_signers));
+        } else {
+            Self::link_account_key_to_did(&new_primary_key, target_did)
+        }
 
         // Replace primary key of the owner that initiated key rotation.
-        Self::unlink_account_key_from_did(&old_primary_key, rotation_for_did);
-        record.primary_key = sender.clone();
-        Self::link_account_key_to_did(&sender, rotation_for_did);
-        <DidRecords<T>>::insert(&rotation_for_did, record);
-
+        record.primary_key = new_primary_key.clone();
         Self::deposit_event(RawEvent::PrimaryKeyUpdated(
-            rotation_for_did,
-            old_primary_key,
-            sender,
+            target_did,
+            old_primary_key.clone(),
+            new_primary_key,
         ));
+
+        if let Some(perms) = new_permissions {
```

Comment:
I might be missing something here, but using `new_permissions` as a marker for whether or not the old primary should be added as a secondary seems a bit wonky. How would you express that you want the old primary added as a new primary with no permissions?

Reply:
Adding the old primary key as a secondary key with no permission: `Some(Permissions::empty())`
We must always provide a `Permissions` value when adding a secondary key even for the "no permissions" case.

Reply:
OK - I see how this is used below - we should update the comment for this function anyway as it is currently inaccurate.

Reply:
I had asked to move all the logic into `common_rotate_primary_key` to avoid the heavy cost of loading/storing the DidRecord value multiple times.

Reply:
Makes sense, but the comment:
```
    // Sets the new primariy key, optionally removes it as a secondary key if it is one,
    // and optionally checks the cdd auth.
```
should be updated to reflect the new functionality (and fix typo in "primariy").

Reply:
We should probably also make it clearer that the new primary key must either be:
- unlinked or;
- linked as a secondary key to the target_did

i.e. this function doesn't allow rotating into a secondary key which is linked to a non-target_did identity.